diff --git a/cf/src/connection.js b/cf/src/connection.js
index 8e79170aeb13efe8c3c77ccede0cd3a115b5e1b5..2306a60e4d57d320692e86116dfb8febb3a97d4c 100644
--- a/cf/src/connection.js
+++ b/cf/src/connection.js
@@ -175,8 +175,9 @@ function Connection(options, queues = {}, { onopen = noop, onend = noop, onclose
       return write(toBuffer(q))
         && !q.describeFirst
         && !q.cursorFn
-        && sent.length < max_pipeline
+        // Reserve transactions even when pipelining is disabled.
         && (!q.options.onexecute || q.options.onexecute(connection))
+        && sent.length < max_pipeline
     } catch (error) {
       sent.length === 0 && write(Sync)
       errored(error)
diff --git a/cjs/src/connection.js b/cjs/src/connection.js
index 07f6716702ac888215c86ad6e0071aaf55ae519c..dc2b547b3945aa1e9a58b0849bcaee7371eb644d 100644
--- a/cjs/src/connection.js
+++ b/cjs/src/connection.js
@@ -173,8 +173,9 @@ function Connection(options, queues = {}, { onopen = noop, onend = noop, onclose
       return write(toBuffer(q))
         && !q.describeFirst
         && !q.cursorFn
-        && sent.length < max_pipeline
+        // Reserve transactions even when pipelining is disabled.
         && (!q.options.onexecute || q.options.onexecute(connection))
+        && sent.length < max_pipeline
     } catch (error) {
       sent.length === 0 && write(Sync)
       errored(error)
diff --git a/src/connection.js b/src/connection.js
index 1b1cccde43b4570d5d071d6ffaab2669b3c2065a..5eb3518bd84d2d735c9c03f7ab4065b5354761e7 100644
--- a/src/connection.js
+++ b/src/connection.js
@@ -173,8 +173,9 @@ function Connection(options, queues = {}, { onopen = noop, onend = noop, onclose
       return write(toBuffer(q))
         && !q.describeFirst
         && !q.cursorFn
-        && sent.length < max_pipeline
+        // Reserve transactions even when pipelining is disabled.
         && (!q.options.onexecute || q.options.onexecute(connection))
+        && sent.length < max_pipeline
     } catch (error) {
       sent.length === 0 && write(Sync)
       errored(error)
